### PR TITLE
feat: Add Rich Progress Bar

### LINKED ISSUE #8255: Implement Rich ProgressBar
## 🚀 Feature

Rich ProgressBar Integration

> Rich is a Python library for rich text and beautiful formatting in the terminal.

Github: https://github.com/willmcgugan/rich
Related PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/5500

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Screen Shot 2021-08-17 at 10 11 50 PM

Next PR will be to support rich Model Summary

Fixes #8255 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@kaushikb11 Great work. I tried it out. 

I have a few high level comments:

- I'm seeing some flickering in the terminal because the description of the bar changes from Epoch X to Validation. Is it possible to reserve enough column space in advance to avoid it? 
- Is it possible to not leave the previous progress bar behind, or is this something we want to change now with this progress bar?
- Do you think the trainer should automatically select the rich bar if rich is installed? 
- I think we should not add rich logging to the model summary in this PR and instead split it off.

File: pytorch_lightning/callbacks/progress/__init__.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# limitations under the License.
+"""
+Progress Bars
+=============
+
+Use or override one of the progress bar callbacks.
+
+"""
+from pytorch_lightning.callbacks.progress.base import ProgressBarBase  # noqa: F401
+from pytorch_lightning.callbacks.progress.progress import ProgressBar, tqdm  # noqa: F401
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/__init__.py:22]: I added it here as we are using it in the tests https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/callbacks/test_progress_bar.py#L27

But I could also extend the import.

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    "train": "red",
+    "sanity_check": "yellow",
+    "validate": "yellow",
+    "test": "yellow",
+    "predict": "yellow",
+}
+
+
+class RichProgressBar(ProgressBarBase):
+    def __init__(self, refresh_rate: int = 1):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/rich_progress.py:84]:
```suggestion
    def __init__(self, refresh_rate: float = 1.0):
```
it can be a float, right? for example 0.5 updates only every other second

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        return not self.is_enabled
+
+    def disable(self) -> None:
+        self._enabled = False
+
+    def enable(self) -> None:
+        self._enabled = True
+
+    def setup(self, trainer, pl_module, stage):
+        self.progress = Progress(
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/rich_progress.py:120]: maybe pass arguments with keywords?

**Comment by kaushikb11** [comment]
@awaelchli Thanks for the feedback!

> I'm seeing some flickering in the terminal because the description of the bar changes from Epoch X to Validation. Is it possible to reserve enough column space in advance to avoid it?

Could you share a gif of it happening? 

> Is it possible to not leave the previous progress bar behind, or is this something we want to change now with this progress bar?

Yess! I will also be adding arguments to customize the bar such as `display_per_epoch`, `show_percent`, `custom_metrics_column`, etc in follow-up PRs.

> Do you think the trainer should automatically select the rich bar if rich is installed?

If we support this, I am thinking about what would be the best way for the User to request to switch to `tqdm`.


> I think we should not add rich logging to the model summary in this PR and instead split it off.

Doable! Will do.

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:120]: `columns` are positional Ref: https://github.com/willmcgugan/rich/blob/master/rich/progress.py#L576

**Comment by awaelchli** [comment]
> Could you share a gif of it happening?

progress flicker



the gif refresh rate is too low to show it. it happens when epochs are short and the text for Epoch and Validation don't have the same length, so the whole column changes. 



```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks.progress import RichProgressBar


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        callbacks=RichProgressBar(),
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)


if __name__ == "__main__":
    run()
```

**Comment by kaushikb11** [comment]
> the text for Epoch and Validation don't have the same length, so the whole column changes.

Ahh, I get your point now! Will figure something out. Flickering of columns could be annoying.

**Comment by kaushikb11** [comment]
@awaelchli It should work now!

File: pytorch_lightning/utilities/model_summary.py [hunk 4]
```diff
@@ -435,7 +436,9 @@ def _is_lazy_weight_tensor(p: Tensor) -> bool:
 
 
 def summarize(
-    lightning_module: "pl.LightningModule", mode: Optional[str] = "top", max_depth: Optional[int] = None
+    lightning_module: "pl.LightningModule",
+    mode: Optional[str] = "top",
+    max_depth: Optional[int] = None,
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_summary.py:441]: revert changes here to preserve git blame

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+        total_batches = total_train_batches + self._total_val_batches
+
+        train_description = self._get_train_description(trainer.current_epoch)
+
+        self.main_progress_bar = self.progress.add_task(
+            f"[{STYLES['train']}]{train_description}",
+            total=total_batches,
+        )
+
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py]: There's a minor change with `total_val_batches`

File: pytorch_lightning/callbacks/progress/progress.py [hunk 6]
```diff
@@ -60,203 +52,29 @@ def format_num(n) -> str:
...
     - **validation progress:** only visible during validation;
       shows total progress over all validation datasets.
     - **test progress:** only active when testing; shows total progress over all test datasets.
-
     For infinite datasets, the progress bar never ends.
-
     If you want to customize the default ``tqdm`` progress bars used by Lightning, you can override
     specific methods of the callback class and pass your custom implementation to the
     :class:`~pytorch_lightning.trainer.trainer.Trainer`:
-
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/progress.py:247]: where is this coming from? your auto-formatter definitely does not do the right thing, these spaces are necessary. @tchaton has the same problem. consider turning it off.

File: pytorch_lightning/callbacks/progress/progress.py [hunk 7]
```diff
@@ -271,7 +89,6 @@ def init_validation_tqdm(self):
             together. This corresponds to
             :paramref:`~pytorch_lightning.trainer.trainer.Trainer.process_position` in the
             :class:`~pytorch_lightning.trainer.trainer.Trainer`.
-
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/progress.py:274]: this is the only line that can be removed

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    "predict": "yellow",
+}
+
+
+class RichProgressBar(ProgressBarBase):
+    def __init__(self, refresh_rate: float = 1.0):
+        if not _RICH_AVAILABLE:
+            raise MisconfigurationException(
+                "`RichProgressBar` requires `rich` to be installed. Install it by running `pip install rich`."
+            )
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/rich_progress.py:88]:
```suggestion
    def __init__(self, refresh_rate: float = 1.0):
        """Create a progress bar with `rich text formatting <https://github.com/willmcgugan/rich>`_.
        
        Args:
            refresh_rate: the number of updates per second, must be strictly positive
        """
        if not _RICH_AVAILABLE:
            raise MisconfigurationException(
                "`RichProgressBar` requires `rich` to be installed. Install it by running `pip install rich`."
            )
```

**Review by awaelchli — APPROVED** [review]
this looks great as a first version
for my personal taste it is too colorful, there is no meaning to the colors and I believe a simple dual-color branding would look much cleaner.

also, expect there to be a couple display issues we have to fix in the future.

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/progress/rich_progress.py:88]: should also add the note to the docstring that we should pip install rich

Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py]: Updated!

File: pytorch_lightning/callbacks/progress/progress.py (hunk 6 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/progress.py]: Resolved! Thanks for the catch ^^

**Comment by kaushikb11** [comment]
> for my personal taste it is too colorful, there is no meaning to the colors and I believe a simple dual-color branding would look much cleaner.

That's the reason I didn't add colors to the metrics. We could define meaning to the colors and the design with one more iteration. :) 

We could support dual-color branding with an extra argument to the Callback. The styles and colors are configurable :D

**Comment by SeanNaren** [comment]
> > for my personal taste it is too colorful, there is no meaning to the colors and I believe a simple dual-color branding would look much cleaner.
> 
> That's the reason I didn't add colors to the metrics. We could define meaning to the colors and the design with one more iteration. :)
> 
> We could support dual-color branding with an extra argument to the Callback. The styles and colors are configurable :D

Agree with Adrian here, maybe we can try a few iterations on colour schemes to see what sticks! 

I'm hoping we can intuitively add some stats to the bar that would be neat, such as average GPU mem allocation. It might also be worth trying to compact the progress bar a bit to reduce space, but a personal preference :)

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by ananthsub** [comment]
@kaushikb11 on naming, would we rename the existing progress bar as TQDMProgressBar? To clarify there's no code dependency between the rich one and the existing one?

**Comment by carmocca** [comment]
Yeah, have the following

```python
class TQDMProgressBar(ProgressBarBase):
    ...

# points to the default implementation
ProgressBar = TQDMProgressBar
```

**Comment by daniellepintz** [comment]
@ananthsub @carmocca I can rename the existing one after #8985

Also @kaushikb11 I am wondering if you forgot to add the RichProgressBar to the documentation, i.e. callbacks.rst?

File: pytorch_lightning/utilities/imports.py [hunk 9]
```diff
@@ -84,6 +84,7 @@ def _compare_version(package: str, op, version) -> bool:
 _NATIVE_AMP_AVAILABLE = _module_available("torch.cuda.amp") and hasattr(torch.cuda.amp, "autocast")
 _OMEGACONF_AVAILABLE = _module_available("omegaconf")
 _POPTORCH_AVAILABLE = _module_available("poptorch")
+_RICH_AVAILABLE = _module_available("rich")
```
Comment (ananthsub) [inline pytorch_lightning/utilities/imports.py:87]: @kaushikb11 - is there a version of rich we should specify here that's compatible with the progress bar features?
